File: Store/src/Sales.Domain/Model/Orders/Order.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using Sales.Domain.Model.Discounts;

namespace Sales.Domain.Model.Orders
{
    public class Order
    {
        private List<OrderItem> _items;
        public long Id { get; private set; }
        public AppliedDiscount AppliedDiscount { get; private set; }
        public IReadOnlyList<OrderItem> Items => _items.AsReadOnly();
        public Order(long id, List<OrderItem> items)
        {
            this.Id = id;
            this._items = items;
        }
        public long TotalPrice()
        {
            var totalPrice = Items.Sum(a => a.TotalPrice());
            if (this.AppliedDiscount == null) return totalPrice;
            if (this.AppliedDiscount.Value > totalPrice) return 0;
            return totalPrice - AppliedDiscount.Value;
        }
        public void ApplyDiscount(Discount discount)
        {
            if (discount.ExpirationTime < DateTime.Now) return;
            var discountValue = discount.CalculateDiscountFor(this);
            this.AppliedDiscount = new AppliedDiscount(discount.Id, discountValue);
        }
    }
}

File: src/main/java/store/sales/domain/model/orders/Order.java
package store.sales.domain.model.orders;

import store.sales.domain.model.discounts.Discount;

import java.util.List;

public class Order {
    private final List<OrderItem> items;
    private final long id;
    private AppliedDiscount appliedDiscount;

    public Order(long id, List<OrderItem> items) {
        this.id = id;
        this.items = items;
    }

    public int totalPrice() {
        int totalPrice = 0;
        for (OrderItem item : items) {
            totalPrice += item.totalPrice();
        }
        if (this.appliedDiscount == null) return totalPrice;
        //if (this.appliedDiscount.getValue() > totalPrice) return 0;
        return totalPrice;
    }

    public void applyDiscount(Discount discount) {
        var discountValue = discount.calculateDiscountFor(this);
        this.appliedDiscount = new AppliedDiscount(discount.getId(), discountValue);
    }

    public long getId() {
        return id;
    }

    public List<OrderItem> getItems() {
        return items;
    }

    public AppliedDiscount getAppliedDiscount() {
        return appliedDiscount;
    }
}
